src/inspector_io.cc

```diff
@@ -62,402 +56,266 @@ std::string GenerateID() {
-}
+  void Wait() {
+    Mutex::ScopedLock scoped_lock(state_lock_);
+    if (messages_.empty()) {
+      incoming_message_cond_.Wait(scoped_lock);
+    }
+  }
 
+  void SetServer(InspectorSocketServer* server) {
+    server_ = server;
+  }
+
+ private:
+  ~RequestQueue() {
+  }
+
+  static void UvAsyncCb(uv_async_t* async) {
+    RequestQueue* wrapper =
+        node::ContainerOf(&RequestQueue::async_, async);
+    wrapper->DoDispatch();
+  }
+
+  static void FreeOnClose(uv_handle_t* handle) {
+    RequestQueue* wrapper =
+        node::ContainerOf(&RequestQueue::async_,
+                          reinterpret_cast<uv_async_t*>(handle));
+    delete wrapper;
+  }
+
+  MessageQueue GetMessages() {
+    Mutex::ScopedLock scoped_lock(state_lock_);
+    MessageQueue messages;
+    messages_.swap(messages);
+    return messages;
+  }
+
+  void DoDispatch() {
+    if (server_ == nullptr)
+      return;
+    for (const auto& outgoing : GetMessages()) {
+      int session_id = std::get<1>(outgoing);
+      switch (std::get<0>(outgoing)) {
+      case TransportAction::kKill:
+        server_->TerminateConnections();
+        // Fallthrough
+      case TransportAction::kStop:
+        if (!stopped_) {
+          server_->Stop();
+          stopped_ = true;
+        }
+        break;
+      case TransportAction::kSendMessage:
+        server_->Send(session_id,
+                      protocol::StringUtil::StringViewToUtf8(
+                          std::get<2>(outgoing)->string()));
+        break;
+      }
+    }
+  }
+
+  uv_async_t async_;
+  InspectorSocketServer* server_ = nullptr;
+  MessageQueue messages_;
+  Mutex state_lock_;  // Locked before mutating the queue.
+  ConditionVariable incoming_message_cond_;
+  bool stopped_ = false;
+};
 
 class IoSessionDelegate : public InspectorSessionDelegate {
  public:
-  explicit IoSessionDelegate(InspectorIo* io, int id) : io_(io), id_(id) { }
-  void SendMessageToFrontend(const v8_inspector::StringView& message) override;
+  explicit IoSessionDelegate(
+      std::shared_ptr<RequestQueue> queue,
+      int id)
+      : request_queue_(queue), id_(id) { }
+  ~IoSessionDelegate() {
+  }
+  void SendMessageToFrontend(const v8_inspector::StringView& message) override {
+  auto queue = request_queue_.lock();
+  if (queue)
+    queue->Post(id_, TransportAction::kSendMessage,
+                StringBuffer::create(message));
+  }
+  bool WaitForDisconnectOnShutdown() override {
+    return true;
+  }
+
  private:
-  InspectorIo* io_;
+  std::weak_ptr<RequestQueue> request_queue_;
   int id_;
 };
 
 // Passed to InspectorSocketServer to handle WS inspector protocol events,
 // mostly session start, message received, and session end.
 class InspectorIoDelegate: public node::inspector::SocketServerDelegate {
  public:
-  InspectorIoDelegate(InspectorIo* io, const std::string& target_id,
+  InspectorIoDelegate(std::shared_ptr<RequestQueue> queue,
+                      std::weak_ptr<MainThreadInterface> interface,
```

Comment:
Looks like this is a first usage of std::weak_ptr in Node codebase. Standard seems to require it to be thread-safe. I would like to confirm with people familiar with less wide-spread platforms it is ok to use it here.